### PR TITLE
Ft reset password via email 182201722

### PR DESCRIPTION
**Pull request by HiHeritier**
**What does this PR do?**

feature to reset password

**Description of Task to be completed?**

- User can be able to reset password

**How should this be manually tested?**

- Clone the repo checkout to branch `ft-reset-password-via-email-182201722`
- Run `npm install` to install modules
- Run `npm run start:prod` to start the production server
- In your API testing tool (postman recommended or swagger doc(`http://localhost:8000/api/v1/docs/`) ) call api `localhost:8000/api/v1/user/forgotpassword` to have the token to reset password sent to your email
-Copy the token sent to your email then call `api localhost:8000/api/v1/user/resetpassword/$token`  to inter new password

**What are the relevant pivotal tracker stories?**
https://www.pivotaltracker.com/story/show/182201722

**Screenshots**
Screenshot (196)

Screenshot (197)

Screenshot (198)

### CODE REVIEW

File: src/controllers/authentication.js [hunk 1]
```diff
@@ -180,3 +183,63 @@ export const logout = (req, res) => {
...
+  
+    const resetURL = `${req.protocol}://${req.get(
+      'host'
+    )}/api/v1/user/resetpassword/${resetToken}`;
+  
+    try {
+      await new Email(verifiedUser, resetURL).sendPasswordReset();
+      res.status(200).json({
+        status: 'success',
+        message: 'token send to Email',
```
Comment (joshwambere) [inline src/controllers/authentication.js]: can you rerun a message not too technical, average user wont know what's a token

File: src/controllers/authentication.js [hunk 2]
```diff
@@ -180,3 +183,63 @@ export const logout = (req, res) => {
...
+      });
+    } catch (err) {  
+        console.log(err)
```
Comment (joshwambere) [inline src/controllers/authentication.js]: Remove this console

File: src/controllers/authentication.js [hunk 3]
```diff
@@ -180,3 +183,63 @@ export const logout = (req, res) => {
...
+        new AppError('There was an error sending email. Try again later.', 500)
+      );
+    }
+  });
+  
+  export const resetPassword = catchAsync(async (req, res, next) => {
+    const hashedToken = createHash('sha256')
+      .update(req.params.token)
+      .digest('hex');
+
```
Comment (joshwambere) [inline src/controllers/authentication.js]: this section needs some linting

File: src/helpers/validation_schema.js [hunk 4]
```diff
@@ -15,7 +15,7 @@ export const signupAuthSchema = Joi.object({
 
     birthDate: Joi.date(),
 
-    phoneNumber: Joi.string().regex(/^[0-9]{10}$/).messages({'string.pattern.base': `Phone number must have 10 digits.`}).required(),
+    phoneNumber: Joi.string().regex(/^[0-9]{10}$/).messages({'string.pattern.base': `Phone number must have 10 digits.`}),
```
Comment (joshwambere) [inline src/helpers/validation_schema.js:12]: why did you remove required on phoneNumber field

**Review by joshwambere — CHANGES_REQUESTED** [review]

**Comment by joshwambere** [comment]
and test on your branch are failing you might want check why

**Review by Patricknshimiyimana — DISMISSED** [review]
Looks good to me

File: src/controllers/authentication.js [hunk 5]
```diff
@@ -180,3 +183,61 @@ export const logout = (req, res) => {
...
+    )}/api/v1/user/resetpassword/${resetToken}`;
+  
+    try {
+      await new Email(verifiedUser, resetURL).sendPasswordReset();
+      res.status(200).json({
+        status: 'success',
+        message: 'Password Reset sent to your Email',
+      });
+    } catch (err) {  
+        console.log(err)
```
Comment (joshwambere) [inline src/controllers/authentication.js]: delete this console

File: src/routers/api/users.js [hunk 6]
```diff
@@ -24,6 +24,8 @@ userRouter.post('/signup',signup);
 userRouter.post('/login',login);
 userRouter.patch('/update',protect,uploads.single("image"),updateUserProfile);
 userRouter.get('/',getAllUsers)
+userRouter.post('/forgotpassword',forgotPassword)
+userRouter.patch('/resetpassword/:token',resetPassword)
```
Comment (joshwambere) [inline src/routers/api/users.js:31]: can you add some middleware to check new password validity before

**Review by joshwambere — CHANGES_REQUESTED** [review]

**Review by mwisemarierose — DISMISSED** [review]

**Pull request closed without merging**